test/extensions/filters/http/wasm/config_test.cc

```diff
@@ -207,6 +211,176 @@ TEST_P(WasmFilterConfigTest, YamlLoadFromRemoteWASM) {
   cb(filter_callback);
 }
 
+TEST_P(WasmFilterConfigTest, YamlLoadFromRemoteWASMFailOnUncachedThenSucceed) {
+  Envoy::Extensions::Common::Wasm::clearCodeCacheForTesting(true);
+  const std::string code = TestEnvironment::readFileToStringForTest(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/filters/http/wasm/test_data/headers_cpp.wasm"));
+  const std::string sha256 = Hex::encode(
+      Envoy::Common::Crypto::UtilitySingleton::get().getSha256Digest(Buffer::OwnedImpl(code)));
+  const std::string yaml = TestEnvironment::substitute(absl::StrCat(R"EOF(
+  config:
+    vm_config:
+      runtime: "envoy.wasm.runtime.)EOF",
+                                                                    GetParam(), R"EOF("
+      code:
+        remote:
+          http_uri:
+            uri: https://example.com/data
+            cluster: cluster_1
+            timeout: 5s
+          sha256: )EOF",
+                                                                    sha256));
+  envoy::extensions::filters::http::wasm::v3::Wasm proto_config;
+  TestUtility::loadFromYaml(yaml, proto_config);
+  WasmFilterConfig factory;
+  NiceMock<Http::MockAsyncClient> client;
+  NiceMock<Http::MockAsyncClientRequest> request(&client);
+
+  EXPECT_CALL(cluster_manager_, httpAsyncClientForCluster("cluster_1"))
+      .WillOnce(ReturnRef(cluster_manager_.async_client_));
+  EXPECT_CALL(cluster_manager_.async_client_, send_(_, _, _))
+      .WillOnce(
+          Invoke([&](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& callbacks,
+                     const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
+            Http::ResponseMessagePtr response(
+                new Http::ResponseMessageImpl(Http::ResponseHeaderMapPtr{
+                    new Http::TestResponseHeaderMapImpl{{":status", "200"}}}));
+            response->body() = std::make_unique<Buffer::OwnedImpl>(code);
+            callbacks.onSuccess(request, std::move(response));
+            return &request;
+          }));
+
+  EXPECT_THROW_WITH_MESSAGE(factory.createFilterFactoryFromProto(proto_config, "stats", context_),
+                            Extensions::Common::Wasm::WasmException,
+                            "Failed to load WASM code (fetching) from https://example.com/data");
+  EXPECT_CALL(init_watcher_, ready());
+  context_.initManager().initialize(init_watcher_);
+  EXPECT_EQ(context_.initManager().state(), Init::Manager::State::Initialized);
+
+  Init::ManagerImpl init_manager2{"init_manager2"};
+  Init::ExpectableWatcherImpl init_watcher2;
+
+  EXPECT_CALL(context_, initManager()).WillRepeatedly(ReturnRef(init_manager2));
+
+  Http::FilterFactoryCb cb = factory.createFilterFactoryFromProto(proto_config, "stats", context_);
```

Comment:
Q: how is time being managed here? 
I see 2 calls to `createFilterFactoryFromProto` with different expectations, but how do we guarantee that 1st fails and 2nd succeeds? The asyncclient can return after 2nd call, no?

Reply:
The way that gtest works is that expectations are dispatched based on the closest one above (before in the code path) which is still active.  In this case I add the ones I am using above the case I am testing.  I am also using Repeatedly so that it is clear which one is going to get selected.  I can add some comments, but it will require rerunning the tests which can take a very long time.